File: 12-insert-delete-getRandom-o(1)/answer.go
package main

import (
	"math/rand"
)

type RandomizedSet struct {
	nums  []int
	index map[int]int // value -> index in nums
}

func Constructor() RandomizedSet {
	return RandomizedSet{
		nums:  []int{},
		index: make(map[int]int),
	}
}

func (this *RandomizedSet) Insert(val int) bool {
	if _, exists := this.index[val]; exists {
		return false
	}
	this.nums = append(this.nums, val)
	this.index[val] = len(this.nums) - 1
	return true
}

func (this *RandomizedSet) Remove(val int) bool {
	idx, exists := this.index[val]
	if !exists {
		return false
	}
	lastVal := this.nums[len(this.nums)-1]
	this.nums[idx] = lastVal
	this.index[lastVal] = idx
	this.nums = this.nums[:len(this.nums)-1]
	delete(this.index, val)
	return true
}

func (this *RandomizedSet) GetRandom() int {
	return this.nums[rand.Intn(len(this.nums))]
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */

File: 12-insert-delete-getRandom-o(1)/answer.py
import random

class RandomizedSet:
    def __init__(self):
        self.nums = []        # stores values
        self.index = {}       # value -> index in nums

    def insert(self, val: int) -> bool:
        if val in self.index:
            return False
        self.nums.append(val)
        self.index[val] = len(self.nums) - 1
        return True

    def remove(self, val: int) -> bool:
        if val not in self.index:
            return False
        idx = self.index[val]
        last_val = self.nums[-1]
        self.nums[idx] = last_val
        self.index[last_val] = idx
        self.nums.pop()
        del self.index[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self.nums)

# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()